### PR TITLE
Cache ee.FeatureCollection promises

### PR DESCRIPTION
**Pull request by janakdr**
Removes any penalty from using the same asset in multiple layers.

### CODE REVIEW

**Comment by janakdr** [comment]
Needed to work around unpkg.com erroring out with @latest deck version. Hopefully will be fixed soon, but might as well pin for now, we're getting close to point where we should pin/download locally.

File: docs/ee_promise_cache.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * Caches results, so there will be only one {@link Promise} per asset path,
+ * avoiding duplicate work in the case that the same asset is requested multiple
+ * times.
+ *
+ * @param {string} eeAssetPath Path to an {@link ee.FeatureCollection}
+ * @return {Promise<GeoJson>}
+ */
+function getEePromiseForFeatureCollection(eeAssetPath) {
+  const maybePromise = cache.get(eeAssetPath);
+  console.log(eeAssetPath, maybePromise);
```
Comment (juliexxia) [inline docs/ee_promise_cache.js]: debugging?

File: docs/ee_promise_cache.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        return;
+      }
+      resolve(resolvedObject);
+    });
+  });
+}
+
+/** Clear cache, for use with tests. */
+function clearForTesting() {
+  cache.clear();
```
Comment (juliexxia) [inline docs/ee_promise_cache.js]: nit: could you rename this clearAssetPromiseCacheForTesting or something like that? Helpful to know what we're clearing.

File: docs/import/manage_disaster.js [hunk 3]
```diff
@@ -1,7 +1,9 @@
 import {eeLegacyPathPrefix, legacyStateDir} from '../ee_paths.js';
+import {convertEeObjectToPromise} from '../ee_promise_cache.js';
 import {disasterCollectionReference} from '../firestore_document.js';
-import {convertEeObjectToPromise, latLngToGeoPoint, transformGeoPointArrayToLatLng} from '../map_util.js';
+import {latLngToGeoPoint, transformGeoPointArrayToLatLng} from '../map_util.js';
 import {getDisaster} from '../resources.js';
+
```
Comment (juliexxia) [inline docs/import/manage_disaster.js]: super nit: extra line

File: docs/run.js [hunk 4]
```diff
@@ -31,30 +28,37 @@ const scalingFactor = 100;
...
 /**
- * Sets {@link scorePromise} to the {@link ee.FeatureCollection} with
- * id {@link getScoreAssetPath}, or, if that does not exist, with id
- * {@link getBackupScoreAssetPath}.  Sets {@link resolvedScoreAsset} to the id
- * of the found collection.
+ * Seeds {@link Promise} cache with `eeAsset`, and returns {@link Promise} whose
+ * value is `eeAsset`. Caller can then wait on returned {@link Promise} to see
+ * if `eeAsset` is valid asset.
+ * @param {string} eeAsset
+ * @return {Promise<string>} {@link Promise} with `eeAsset` once {@link
```
Comment (juliexxia) [inline docs/run.js]: super nit: this {@link Promise} isn't highlighting correctly in github - is it in intellij? if not there either maybe just take out

File: docs/run.js [hunk 5]
```diff
@@ -31,30 +28,37 @@ const scalingFactor = 100;
 let resolvedScoreAsset;
 
 /**
- * Sets {@link scorePromise} to the {@link ee.FeatureCollection} with
- * id {@link getScoreAssetPath}, or, if that does not exist, with id
- * {@link getBackupScoreAssetPath}.  Sets {@link resolvedScoreAsset} to the id
- * of the found collection.
+ * Seeds {@link Promise} cache with `eeAsset`, and returns {@link Promise} whose
```
Comment (juliexxia) [inline docs/run.js]: super nit: could do {@code eeAsset} to make more clear it's not just the string 'eeAsset' (already pretty clear just took a second read through)

File: docs/run.js [hunk 6]
```diff
@@ -31,30 +28,37 @@ const scalingFactor = 100;
...
+ * @param {string} eeAsset
+ * @return {Promise<string>} {@link Promise} with `eeAsset` once {@link
+ *     ee.FeatureCollection} with path `eeAsset` has been downloaded
+ */
+function createPromiseWithPathIfSuccessful(eeAsset) {
+  return getEePromiseForFeatureCollection(eeAsset).then(() => eeAsset);
+}
+
+/**
+ * Sets {@link resolvedScoreAsset} to {@link getScoreAssetPath}, or, if that
```
Comment (juliexxia) [inline docs/run.js]: super nit: I \*\*think\*\* @link is for methods/classes and @code is for variables? At least that's how I've always used them, would be interested to hear if that's not how you use them
https://docs.oracle.com/javase/8/docs/technotes/tools/windows/javadoc.html

File: docs/run.js [hunk 7]
```diff
@@ -31,30 +28,37 @@ const scalingFactor = 100;
...
+ * does not exist as an EarthEngine asset, to {@link getBackupScoreAssetPath}.
```
Comment (juliexxia) [inline docs/run.js]: And if backup doesn't exist?

File: cypress/integration/unit_tests/click_feature_test.js [hunk 8]
```diff
@@ -94,7 +84,8 @@ describe('Unit tests for click_feature.js with map and table', () => {
       tableDiv.id = 'table';
       containerDiv.appendChild(tableDiv);
       drawTableAndSetUpHandlers(
-          convertEeObjectToPromise(scoredFeatures).then((fc) => fc.features),
+          EePromiseCache.convertEeObjectToPromise(scoredFeatures)
```
Comment (juliexxia) [inline cypress/integration/unit_tests/click_feature_test.js]: why not just regular import here? were you stubbing something before?

File: docs/run.js (hunk 4 above)
Comment (janakdr) [reply docs/run.js]: Highlights same as other places.

File: docs/ee_promise_cache.js (hunk 2 above)
Comment (janakdr) [reply docs/ee_promise_cache.js]: Done.

File: cypress/integration/unit_tests/click_feature_test.js (hunk 8 above)
Comment (janakdr) [reply cypress/integration/unit_tests/click_feature_test.js]: Oops, good guess, I was. Removed.

**Pull request merged**